### PR TITLE
Issue #10563: MissedPersistentTimerAction

### PR DESCRIPTION
**Pull request by tkburroughs**
Added support for skipping missed EJB persistent timer expirations

for #10563

### CODE REVIEW

**Comment by tkburroughs** [comment]
#build

**Comment by andrea-pichler** [comment]
No message review from L2 requested

**Comment by tkburroughs** [comment]
#run-libby-bot

**Comment by tkburroughs** [comment]
#build

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -36,3 +36,9 @@ persistentExecutorRef.desc=Schedules and runs EJB persistent timer tasks.
 nonPersistentContextServiceRef=Non-persistent Timer Thread Context Propagation
 nonPersistentContextServiceRef$Ref=Thread context propagation reference
 nonPersistentContextServiceRef.desc=The context service is used to manage context propagation to non-persistent timer method threads.
+
+missedPersistentTimerAction=Missed persistent timer action
+missedPersistentTimerAction.desc=Specifies the action to perform when the expiration of an interval or schedule based persistent timer is missed. One or more expirations of a persistent timer are classified as missed if the current expiration is scheduled prior to application server start or the next expiration is scheduled prior to the current time. The default action when failover of persistent timers is enabled is ONCE, otherwise the default action is ALL.
```
Comment (Rwalls1) [inline dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: Change "schedule based" to "Change "schedule-based"
Change "scheduled prior to application server start" to "scheduled before the application server start"
Change "next expiration is scheduled prior to the current time" to "next expiration is scheduled before the current time"

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -36,3 +36,9 @@ persistentExecutorRef.desc=Schedules and runs EJB persistent timer tasks.
...
+missedPersistentTimerAction.ALL=The timeout method is invoked immediately for all missed expirations. When multiple expirations have been missed for the same timer, each invocation will occur synchronously until all missed expirations have been processed, then the timer will resume with the next future expiration.
```
Comment (Rwalls1) [inline dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: Change "When multiple expirations have been missed" to "When multiple expirations are missed".
Change "each invocation will occur" to "each invocation occurs".
Change "missed expirations have been processed" to "missed expirations are processed".
Change "timer will resume" to "timer resumes".

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties [hunk 3]
```diff
@@ -36,3 +36,9 @@ persistentExecutorRef.desc=Schedules and runs EJB persistent timer tasks.
...
 nonPersistentContextServiceRef.desc=The context service is used to manage context propagation to non-persistent timer method threads.
+
+missedPersistentTimerAction=Missed persistent timer action
+missedPersistentTimerAction.desc=Specifies the action to perform when the expiration of an interval or schedule based persistent timer is missed. One or more expirations of a persistent timer are classified as missed if the current expiration is scheduled prior to application server start or the next expiration is scheduled prior to the current time. The default action when failover of persistent timers is enabled is ONCE, otherwise the default action is ALL.
+missedPersistentTimerAction.ALL=The timeout method is invoked immediately for all missed expirations. When multiple expirations have been missed for the same timer, each invocation will occur synchronously until all missed expirations have been processed, then the timer will resume with the next future expiration. 
+missedPersistentTimerAction.ONCE=The timeout method is invoked once immediately. All other missed expirations are skipped and the timer will resume with the next future expiration.
```
Comment (Rwalls1) [inline dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: Change "timer will resume" to "timer resumes".

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties [hunk 4]
```diff
@@ -36,3 +36,9 @@ persistentExecutorRef.desc=Schedules and runs EJB persistent timer tasks.
...
+missedPersistentTimerAction.ALL=The timeout method is invoked immediately for all missed expirations. When multiple expirations have been missed for the same timer, each invocation will occur synchronously until all missed expirations have been processed, then the timer will resume with the next future expiration. 
+missedPersistentTimerAction.ONCE=The timeout method is invoked once immediately. All other missed expirations are skipped and the timer will resume with the next future expiration. 
+missedPersistentTimerAction.NONE=No action is taken. All missed expirations that occurred prior to application server start are skipped and the timer will resume with the next future expiration. Missed expirations that occur after application server start are handled with the ONCE missed timer action.
```
Comment (Rwalls1) [inline dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: Change "occurred prior to application server start" to "occurred before application server start".
Change "the timer will resume " to "the timer resumes".

**Comment by tkburroughs** [comment]
#run-libby-bot

**Comment by tkburroughs** [comment]
@Rwalls1 Thanks!  I've made the requested updates.

**Comment by tkburroughs** [comment]
#build

File: dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/EJBPersistentTimerRuntimeImpl.java [hunk 5]
```diff
@@ -118,6 +131,14 @@ protected void modified(ComponentContext cc, Map<String, Object> properties) {
 
     private void updateConfiguration(Map<String, Object> properties) {
 
+        // Read the missedPersistentTimerAction configuration; do not set a default as
+        // that will vary based on whether the PersistentExecutor has failover enabled
+        String missedTimerActionProperty = (String) properties.get(MISSED_TIMER_ACTION);
+        missedTimerAction = (missedTimerActionProperty != null) ? MissedTimerAction.valueOf(missedTimerActionProperty) : null;
```
Comment (olendvcook) [inline dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/EJBPersistentTimerRuntimeImpl.java]: `MissedTimerAction.valueOf(missedTimerActionProperty.toUpperCase())`

^ depending on if kernel lets metatype values be case insensitive.

**Comment by olendvcook** [comment]
Also, while reviewing this I thought of a scenario where a customer might want to have to `missedPersistentTimerAction=NONE` or `missedPersistentTimerAction=ONCE` during server startup, but `missedPersistentTimerAction=ALL` after the server has come up and ran through all the timers fired while the server was down, and then do the ALL scenario for server hiccups.

We would have to rethink some of the design, and if this would be worth it /feasible with persistent timer failover.

**Comment by tkburroughs** [comment]
#run-libby-bot

File: dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/EJBPersistentTimerRuntimeImpl.java (hunk 5 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/EJBPersistentTimerRuntimeImpl.java]: Added 2 new test variations to check for mixed case value (kernel changes to match case in metatype) and bad value (kernel logs warning and provides no value).

**Comment by tkburroughs** [comment]
@olendvcook  While ONCE | NONE only for server start, then ALL after server start could be accomplished, using ALL after server start still has all the same disadvantages of ALL in general. If functionally a timer can be skipped on server start, then it seems it may also be skipped later.  The late timer warning will still be logged, and useful for determining when the server is overloaded. The frequency of the timer running would end up being about the same, except there wouldn't be bursts of the timer running repeatedly. 

Also, the behavior becomes more complex with failover.  When a timer fails over to another server, would it treated as if it were server start, or later.

Rather than introduce more options, making configuration more complex. I think we should stick with the current design until a scenario is encountered that requires more complex configurations. If we were to support more complex configurations, it could be done by just adding new "options" rather then new attributes.  For example, "NONE-ALL" could be added as a value, however it may depend on future needs.

**Comment by tkburroughs** [comment]
#build

**Comment by tkburroughs** [comment]
#build

**Comment by olendvcook** [comment]
@tkburroughs Approved, latest 2 builds have a few failures but all match existing defects, we are in Buildcon 1 though.

**Comment by olendvcook** [comment]
#run-libby-bot

File: dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/fat/src/com/ibm/ws/ejbcontainer/timer/persistent/fat/tests/PersistentTimerCoreTest.java [hunk 6]
```diff
@@ -185,4 +220,214 @@ private static void setLateTimerThresholdConfiguration(Long lateTimerThreshold)
...
+     * </ol>
+     */
+    @Test
+    @Mode(Mode.TestMode.FULL)
+    public void testMissedTimerActionOnceWithFailover() throws Exception {
+        testMissedTimerAction("ONCE", true);
+    }
+
+    /**
+     * Test Persistent Timer missed action "ONCE" behavior when failover has not been enabled. <p>
```
Comment (brideck) [inline dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/fat/src/com/ibm/ws/ejbcontainer/timer/persistent/fat/tests/PersistentTimerCoreTest.java]: Copy/paste error -- comment should read "NONE"

File: dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/test-applications/MissedTimerActionWeb.war/src/com/ibm/ws/ejbcontainer/timer/persistent/missed/web/MissedTimerActionServlet.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+     * <li> Interval timer will skip expirations missed because of a delay.
+     * <li> Timer.getNextTimeout() will return values in the future; skipping missed expirations.
+     * </ol>
+     */
+    public void testMissedTimerActionOnceWithFailover() throws Exception {
+        testMissedTimerAction("testMissedTimerActionOnceWithFailover", "ONCE");
+    }
+
+    /**
+     * Test Persistent Timer missed action "ONCE" behavior when failover has not been enabled. <p>
```
Comment (brideck) [inline dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/test-applications/MissedTimerActionWeb.war/src/com/ibm/ws/ejbcontainer/timer/persistent/missed/web/MissedTimerActionServlet.java]: And again, "NONE"

**Comment by tkburroughs** [comment]
Buildcon is now 3; and failures in last build are known issues.  Merging...

File: dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/fat/src/com/ibm/ws/ejbcontainer/timer/persistent/fat/tests/PersistentTimerCoreTest.java (hunk 6 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/fat/src/com/ibm/ws/ejbcontainer/timer/persistent/fat/tests/PersistentTimerCoreTest.java]: Will correct with next test update that will add server restart variations.

File: dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/test-applications/MissedTimerActionWeb.war/src/com/ibm/ws/ejbcontainer/timer/persistent/missed/web/MissedTimerActionServlet.java (hunk 7 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer.persistent_fat/test-applications/MissedTimerActionWeb.war/src/com/ibm/ws/ejbcontainer/timer/persistent/missed/web/MissedTimerActionServlet.java]: Will correct with next test update that will add server restart variations.

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties (hunk 1 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: done

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties (hunk 2 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: done

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties (hunk 3 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: done

File: dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties (hunk 4 above)
Comment (tkburroughs) [reply dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties]: done

**Pull request merged**